torch/csrc/utils/future.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+
+namespace utils {
+
+// FutuerError inherits from std::exception, it can return const char* or
+// std::string error message
+struct TORCH_API FutureError final : public std::exception {
+  FutureError(std::string&& error_msg_) : error_msg(std::move(error_msg_)) {}
+
+  FutureError() = default;
+
+  const char* what() const noexcept override {
+    return error_msg.c_str();
+  }
+
+  std::string errMsg() const {
+    return error_msg;
+  }
+
+  std::string error_msg;
+};
+
+// This class holds a value of type T that will be ready in the future.
+// Most implementation is copied from FutureMessage and
+// c10::ivalue::Future
+template <typename T>
+class TORCH_API Future final {
+ public:
+  using Callback = std::function<void(const T&, bool, const FutureError&)>;
+
+  const T& wait() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    finished_cv_.wait(lock, [this] { return completed_.load(); });
+
+    if (has_error_) {
+      throw error_;
+    }
+    return value_;
+  }
+
+  void markCompleted(T value) {
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      TORCH_CHECK(!completed());
+      completed_ = true;
+      value_ = std::move(value);
+
+      fireCallbacks();
+    }
+    finished_cv_.notify_all();
+  }
+
+  void markCompleted() {
+    markCompleted(T());
+  }
+
+  void markCompleted(FutureError&& error) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    AT_ASSERT(!completed());
+    completed_ = true;
+    has_error_ = true;
+    error_ = std::move(error);
+
+    fireCallbacks();
+    finished_cv_.notify_all();
+  }
+
+  bool completed() const {
+    return completed_;
+  }
+
+  T value() {
```

Comment:
This is making a copy of the value, when do we plan to use this?

Reply:
this is to make it compatible with c10::ivalue::future, let me remove it for now. actually it will not be used